File: app/src/main/java/hibernate/v2/ringtonerandomizer/ui/adapter/RingtoneSelectAdapter.java
package hibernate.v2.ringtonerandomizer.ui.adapter;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import hibernate.v2.ringtonerandomizer.R;
import hibernate.v2.ringtonerandomizer.model.Ringtone;

/**
 * Created by himphen on 25/5/16.
 */
public class RingtoneSelectAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

	private List<Ringtone> mDataList;
	private ItemClickListener mClickListener;

	public interface ItemClickListener {
		void onItemDetailClick(Ringtone ringtone);
	}

	public RingtoneSelectAdapter(@NonNull List<Ringtone> mDataList,
	                             ItemClickListener mClickListener) {
		this.mDataList = mDataList;
		this.mClickListener = mClickListener;
	}

	@NonNull
	@Override
	public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
		Context context = parent.getContext();

		View itemView = LayoutInflater.from(context).inflate(R.layout.list_item_select, parent, false);
		return new ItemViewHolder(itemView);
	}

	@Override
	public void onBindViewHolder(@NonNull RecyclerView.ViewHolder rawHolder, int position) {
		Ringtone ringtone = mDataList.get(position);
		ItemViewHolder holder = (ItemViewHolder) rawHolder;

		holder.filenameTv.setText(ringtone.getName());
		if (ringtone.getPath() == null) {
			holder.filepathTv.setText("Internal Storage");
		} else {
			holder.filepathTv.setText(ringtone.getPath());
		}

		holder.rootView.setTag(ringtone);
		holder.rootView.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				mClickListener.onItemDetailClick((Ringtone) v.getTag());
			}
		});

		holder.checkbox.setChecked(ringtone.getChecked());
		holder.checkbox.setTag(ringtone);
		holder.checkbox.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				Ringtone ringtone = (Ringtone) v.getTag();
				boolean b = !ringtone.getChecked();
				ringtone.setChecked(b);
				((CheckBox) v).setChecked(b);
			}
		});
	}

	@Override
	public int getItemCount() {
		return mDataList.size();
	}

	public void refreshData(@NonNull List<Ringtone> mDataList) {
		this.mDataList.clear();
		this.mDataList.addAll(mDataList);
		this.notifyDataSetChanged();
	}

	public ArrayList<Ringtone> getCheckedRingtoneList() {
		ArrayList<Ringtone> checkedRingtoneList = new ArrayList<>();
		for (Ringtone ringtone : mDataList) {
			if (ringtone.getChecked()) {
				checkedRingtoneList.add(ringtone);
			}
		}

		return checkedRingtoneList;
	}

	static class ItemViewHolder extends RecyclerView.ViewHolder {

		@BindView(R.id.root_view)
		LinearLayout rootView;
		@BindView(R.id.filenameTv)
		TextView filenameTv;
		@BindView(R.id.filepathTv)
		TextView filepathTv;
		@BindView(R.id.checkbox)
		CheckBox checkbox;

		ItemViewHolder(View itemView) {
			super(itemView);
			ButterKnife.bind(this, itemView);
		}
	}
}

File: app/src/main/java/hibernate/v2/ringtonerandomizer/ui/adapter/RingtoneSelectAdapter.kt
package hibernate.v2.ringtonerandomizer.ui.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import hibernate.v2.ringtonerandomizer.R
import hibernate.v2.ringtonerandomizer.model.Ringtone
import java.util.ArrayList

/**
 * Created by himphen on 25/5/16.
 */
class RingtoneSelectAdapter(
        private val dataList: MutableList<Ringtone>,
        private val clickListener: ItemClickListener
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    interface ItemClickListener {
        fun onItemDetailClick(ringtone: Ringtone)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return ItemViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.list_item_select, parent, false))
    }

    override fun onBindViewHolder(rawHolder: RecyclerView.ViewHolder, position: Int) {
        val ringtone = dataList[position]
        val holder = rawHolder as ItemViewHolder
        holder.filenameTv.text = ringtone.name
        if (ringtone.path == null) {
            holder.filepathTv.text = Ringtone.PATH_INTERNAL_STORAGE
        } else {
            holder.filepathTv.text = ringtone.path
        }
        holder.rootView.tag = ringtone
        holder.rootView.setOnClickListener { v -> clickListener.onItemDetailClick(v.tag as Ringtone) }
        holder.checkbox.isChecked = ringtone.checked
        holder.checkbox.tag = ringtone
        holder.checkbox.setOnClickListener { view ->
            val ringtone1 = view.tag as Ringtone
            val b = !ringtone1.checked
            ringtone1.checked = b
            (view as CheckBox).isChecked = b
        }
    }

    override fun getItemCount(): Int = dataList.size

    fun refreshData(mDataList: List<Ringtone>) {
        this.dataList.clear()
        this.dataList.addAll(mDataList)
        notifyDataSetChanged()
    }

    val checkedRingtoneList: ArrayList<Ringtone>
        get() {
            val checkedRingtoneList = ArrayList<Ringtone>()
            for (ringtone in dataList) {
                if (ringtone.checked) {
                    checkedRingtoneList.add(ringtone)
                }
            }
            return checkedRingtoneList
        }

    internal class ItemViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var rootView: LinearLayout = view.findViewById(R.id.root_view)
        var filenameTv: TextView = view.findViewById(R.id.filenameTv)
        var filepathTv: TextView = view.findViewById(R.id.filepathTv)
        var checkbox: CheckBox = view.findViewById(R.id.checkbox)
    }

}